### ISSUE TITLE
-Woverloaded-shift-op-parentheses bugs

### ISSUE THREAD

**Issue by jeremy-rifkin** [llvm/llvm-project#63897]
**Background:** I am the author of a library called libassert. The library does some tricks to decompose a user-provided expression into a left and right hand side with something along the lines of `expression_decomposer{} << expr`.

**Bug:**
This triggers `-Woverloaded-shift-op-parentheses`
```cpp
class test_class {
    void something_else() {
        std::string s = "test\n";
        ASSERT(s == "test");
    }
};
```
This does not:
```cpp
void something_else() {
    std::string s = "test\n";
    ASSERT(s == "test");
}
```

Additionally, `_Pragma("GCC diagnostic ignored \"-Woverloaded-shift-op-parentheses\"")` doesn't work to suppress the buggy error message. https://godbolt.org/z/ef9sn4fYb

The relevant part of the code in question looks something like

```cpp
 #if LIBASSERT_IS_GCC
  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
     _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
     _Pragma("GCC diagnostic ignored \"-Wuseless-cast\"") // #49
 #else
  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
     _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
     _Pragma("GCC diagnostic ignored \"-Woverloaded-shift-op-parentheses\"")
 #endif
...
#define ASSERT_INVOKE(expr, doreturn, check_expression, name, type, failaction, ...) \
        LIBASSERT_IGNORE_UNUSED_VALUE \
        LIBASSERT_STMTEXPR( \
          LIBASSERT_WARNING_PRAGMA_PUSH \
          LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
          auto libassert_decomposer = \
                         libassert::detail::expression_decomposer(libassert::detail::expression_decomposer{} << expr); \
          LIBASSERT_WARNING_PRAGMA_POP \
          ...
```

https://github.com/jeremy-rifkin/libassert/blob/55d05fb987bea8d6080718dc6262b9cb8a8ad17b/include/assert.hpp

I apologize as I have not yet been able to create a minimal reproducible example yet

**Comment by jeremy-rifkin** [comment]
I'm able to reduce the error at least to this
https://godbolt.org/z/MoM7de31o

Replacing `LIBASSERT_WARNING_PRAGMA_PUSH`/`POP` with no-ops allows the compilation to succeed
```cpp
#define LIBASSERT_WARNING_PRAGMA_PUSH _Pragma("GCC diagnostic push")
#define LIBASSERT_WARNING_PRAGMA_POP  _Pragma("GCC diagnostic pop")
```

It also succeeds without the `IDENT` macro
```cpp
#define ASSERT_INVOKE(expr) \
          LIBASSERT_WARNING_PRAGMA_PUSH \
          LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
          auto libassert_decomposer = \
                         libassert::detail::expression_decomposer(libassert::detail::expression_decomposer{} << expr); \
          LIBASSERT_WARNING_PRAGMA_POP
```

**Comment by jeremy-rifkin** [comment]
This may not be a -Woverloaded-shift-op-parentheses issue of course, too

**Comment by efriedma-quic** [comment]
Reduced further:

```
struct S  {
    S operator<<(S);
    S operator==(const char*);
    void something_else() {
#define MACRO(B) B
        MACRO( 
        _Pragma("GCC diagnostic push")
        _Pragma("GCC diagnostic ignored \"-Woverloaded-shift-op-parentheses\"")
        (void)(S{} << S{} == "test");
        // aaaaaaaaaaaaaaaaaaaaaaaaaaaaa
        _Pragma("GCC diagnostic pop")
        )
    }
};
```

This is probably a generic issue with the interaction between _Pragma and C++ inline member functions.

**Comment by shafik** [comment]
CC @AaronBallman
